### DISCUSSION TITLE
Empty String Effect (vscode editor render effect; parsed result is ok)

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#659] [Q&A] [answered]
I have a (probably stupid) question but I did not find a solution in the discussions (I use langium 0.4.0)... 

I have a grammar which allows string inputs:
```
terminal STRING returns string: /("([^"])*")|('([^'])*')/;
```

But somehow, the vscode editor interprets empty string as the start of a multiline string(?):  
![image](https://user-images.githubusercontent.com/16887130/187248169-d41c4a9c-f42b-40ea-bfe8-2acad80631d5.png)

Is there a solution to that problem?

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

that's probably due to the textMate syntax that we generate. The heuristic that we use probably has generated something unsensible in your case. Can you shared what your `tmLanguage.json` file contains?

Also, I believe that `terminal STRING returns string: /"[^"]*"|'[^']*'/;` (the same as your terminal, just without the `()`) works as expected and actually parses the same way.

**Reply by goto40** [reply]
Here is the `*.tmLanguage.json`:
 * Probably the the `"end": "[^']'"` is not ok - it does not allow an empty string...?
 * Is this file generated? - I did not edit it.
```
{
  "name": "message-dsl",
  "scopeName": "source.message-dsl",
  "fileTypes": [
    ".msg",
    ".item"
  ],
  "patterns": [
    {
      "include": "#comments"
    },
    {
      "name": "keyword.control.message-dsl",
      "match": "\\b(applicable|applicable_for|array|by|constant|constants|description|embedded|enum|enum_value|extends|for|if|import|message|optional|package|per|property|property_set|rawtype|scalar|struct|struct_definition|times|times_per_message|to|value|variant)\\b"
    },
    {
      "name": "string.quoted.double.message-dsl",
      "begin": "\"",
      "end": "[^\"]\""
    },
    {
      "name": "string.quoted.single.message-dsl",
      "begin": "'",
      "end": "[^']'"
    }
  ],
  "repository": {
    "comments": {
      "patterns": [
        {
          "name": "comment.block.message-dsl",
          "begin": "/\\*",
          "beginCaptures": {
            "0": {
              "name": "punctuation.definition.comment.message-dsl"
            }
          },
          "end": "\\*/",
          "endCaptures": {
            "0": {
              "name": "punctuation.definition.comment.message-dsl"
            }
          }
        },
        {
          "begin": "//",
          "beginCaptures": {
            "1": {
              "name": "punctuation.whitespace.comment.leading.message-dsl"
            }
          },
          "end": "(?=$)",
          "name": "comment.line.message-dsl"
        }
      ]
    }
  }
}
```

P.S. You are right, the parentheses are not necessary.

**Reply by msujew** [reply]
Yes the file is generated. The CLI tries to guess what certain terminals mean, how they start and end, and generates the `.tmLanguage.json` based off of that. The file is responsible for syntactic highlighting in vscode.

I believe you can probably strip even more parenthesis, just use `/"[^"]*"|'[^']*'/`. This definitely generates correctly.

**Reply by goto40** [reply]
Thank you, @msujew.
Great hint! Works.
